subsys/net/lib/prometheus/formatter.c

```diff
@@ -0,0 +1,226 @@
+/*
+ * Copyright (c) 2024 Mustafa Abdullah Kus, Sparse Technology
+ *
+ * SPDX-License-Identifier: Apache-2.0
+ */
+
+#include <zephyr/net/prometheus/formatter.h>
+#include <zephyr/net/prometheus/collector.h>
+#include <zephyr/net/prometheus/metric.h>
+#include <zephyr/net/prometheus/histogram.h>
+#include <zephyr/net/prometheus/summary.h>
+#include <zephyr/net/prometheus/gauge.h>
+#include <zephyr/net/prometheus/counter.h>
+
+#include <stdarg.h>
+#include <stdio.h>
+#include <string.h>
+#include <stddef.h>
+
+#include <zephyr/kernel.h>
+
+#include <zephyr/logging/log.h>
+LOG_MODULE_REGISTER(pm_formatter, CONFIG_PROMETHEUS_LOG_LEVEL);
+
+static int write_to_buffer(char *buffer, size_t buffer_size, const char *str)
+{
+	/* helper function to write formatted string to buffer, handling buffer size */
+	int len = snprintf(buffer + strlen(buffer), buffer_size - strlen(buffer), "%s", str);
+
+	if (len >= 0 && len < buffer_size - strlen(buffer)) {
+		return 1;
+	}
+
+	/* error (insufficient buffer size) */
+	return 0;
```

Comment:
Zephyr functions typically `return 0` on success and a negative value on failure, like libc syscalls.

A typical idiom is:
```
int ret;

ret = function_call_that_might_fail();
if (ret < 0) {
        LOG_ERR("optional error message");
        return ret;
}

return 0;
```

`int err;` and `if (err) { return err; };` is another variant used.

Reply:
You would need to propagate this change all the way down:
```diff
-if (!function(args)) {
-        return 0;
-}
+int err;
+
+err = function(args);
+if (err < 0) {
+        return 0;
+}
```

It is a bit more verbose, but a very frequent idiom.

Reply:
Thank you for the effort of changing it everywhere!

Some last minor edit at this line:
```suggestion
	return -ENOBUFS;
```

`NOBUFS` is for "No buffer space available". Maybe `-ENOMEM` or something else is also suited.